### ISSUE TITLE
Bug: Wrong handler mapping after esbuild

### ISSUE THREAD

**Issue by santiperone** [aws/aws-sam-cli#4082]
### Description:
  When running esbuild on a Lambda function, the output doesn't mantain the same folder structure as the enry points, and this remains unchange in the built template, resulting in SAM not finding the lambda Handler:

### Steps to reproduce:
Project structure:
```
- package.json
- template.yaml
-- src
--- functions
---- FunctionName
----- app.js
--- utils
---- some-util.js
```

Template.yaml:
```
FunctionName:
    Type: AWS::Serverless::Function
    Metadata:
      BuildMethod: esbuild
    Properties:
      Handler: src/functions/FunctionName/app.Handler
      Events:
        HttpEvent:
          ....
```

### Observed result:
code is budled to FunctionName folder in .aws-build resulting in:
```
- .aws-build
-- build
--- template.yaml
--- FunctionName
---- app.js
```
However, the built template results in:

```
FunctionName:
    Type: AWS::Serverless::Function
    Metadata:
      BuildMethod: esbuild
    Properties:
      CodeUri: FunctionName
      Handler: src/functions/FunctionName/app.Handler
      Events:
        HttpEvent:
          ....
```

Resulting in an error since FunctionName/src/functions/FunctionName/app.Handler

### Expected result:
Handler in built template should match esbuild output. I.E: 

```
FunctionName:
    Type: AWS::Serverless::Function
    Metadata:
      BuildMethod: esbuild
    Properties:
      CodeUri: FunctionName
      Handler: app.Handler
      Events:
        HttpEvent:
          ....
```

**Comment by jfuss** [comment]
@santiperone You need to provide a `CodeUri` in your template. If you do not, SAM CLI assumes `CodeUri` is the `cwd`. This behavior cannot be changed since it would cause a breaking change. 

Looks like you want the `CodeUri` to be `src/functions/FunctionName` and Handler to be `app.Handler`, from your example above.

**Comment by santiperone** [comment]
@jfuss I am aware of this. Actually no, I need to bundle packages and files from the entire cwd, (i.e. node_modules and the 'some_util' file from the example), and if I set my codeUri as you say this files would not be found by esbuild.

**Comment by santiperone** [comment]
This approach would allow to have shared, and a single node_modules folder (which makes testing a set of lambdas within a project easier) whilst keeping only the needed code for each function upon build, which I guess is the point of using esbuild.

**Comment by moelasmar** [comment]
@santiperone .. what is the point of setting the `Handler` value to `src/functions/FunctionName/app.Handler`. You need to refer to the handler that will be exist in the package that will be deployed to AWS.

**Comment by santiperone** [comment]
@moelasmar Actually that would allow running the code with sam local start-api without the need of previously building it, but you're right, defininf the handler and CodeUri actually works as expected.

**Comment by santiperone** [comment]
@mildaniel I still think this is worth having a look at. It would also allow to have one single codeUri within a project (hence a single build, installing dependendencies only once).

**Comment by lucashuy** [comment]
Hi, thanks for opening this issue! Having root level node_modules is not a case that SAM CLI supports right now. The solution would be to set the CodeUri according to your folder structure and Handler to your functions like mentioned previously.

Please feel free to reopen or open a new issue if anything else arises.

**Comment by santiperone** [comment]
@lucashuy Actually, this is something that is even included in one of SAM's examples https://github.com/aws/aws-sam-cli-app-templates/tree/master/nodejs16.x/cookiecutter-quick-start-web so how come it is not actually supported? Furthermore, this limits us to only having the possibility to have the handler at the root level of the codeUri (You can't have something like ./src/app.js for example, because the output from esbuild would be wrongly mapped) And sharing a single node_modules at root level implies Only one npm install is run for all of your lambda functions, which would make the build process much faster.

**Comment by lucashuy** [comment]
Due to how the build process works, the source directory (the one with the `template.yaml`) is copied to a scratch folder. This includes the source files and the `package.json` file. This means that for each function, `npm install` is run using the same package file. So while the sample project only has one package file, that file is copied for all the sample functions.

Since the discussion is root level node_modules oriented, could you create a new feature request issue regarding root node_modules so that we can gauge community interest?

**Comment by santiperone** [comment]
I see, I was under the impression the build runs once for each CodeUri, at least it seems to do that when you try it out, but I haven't actually taken a look at the code. If that was the case, I guess it wouldn't actually make the builds fuster, but it would still however make the project easier to organize.

**Comment by lucashuy** [comment]
The builder is run per function instead of per CodeUri. In any case, I'd like to move further discussion about this topic to https://github.com/aws/aws-sam-cli/issues/2655 since this is all related to building functions from a single location 😄

**Comment by mildaniel** [comment]
I'm going to re-open this issue to investigate how we can resolve the handler path pointing to the wrong location in the built template.

**Comment by mildaniel** [comment]
I did some digging and found the issue is with how we bundle. I've explained it in #4216, but essentially we bundle straight into the artifact directory, at which point we've already lost the file tree from the code URI to the handler.
e.g. esbuild takes the entry point (`src/functions/FunctionName/app.Handler`) and bundles that into a single file (`app.js`) then takes that file and moves it into the artifact directory `FunctionName`

**Comment by santiperone** [comment]
@mildaniel any news on this?

**Comment by tunesmith** [comment]
I've found a similar problem when using regular `sam init` (v 1.57):

Quick Start, Hello World, nodejs16.x, Zip, Typescript. 

First thing to note is that if I invoke `sam build` and then `sam local start-api -d 3099` and then attach my debugger as I invoke a curl command, I am able to set breakpoints in `hello-world/app.ts` and step through with my debugger.

If I then add two subfolders:

```
- template.yaml
- hello-world
-- package.json
-- app.ts
-- node_modules
-- hello1
--- app.ts
-- hello2
--- app.ts
```

Then I'm faced with a conundrum:

- If, for each `AWS::Serverless::Function`, I point `CodeUri` to a subfolder like `hello1`, it complains about not finding a package.json. It says `package.json file not found. Bundling source without dependencies.` The resultant `app.js.map` in `.aws-sam` does have an accurate `sources` entry though, e.g. `"sources": ["../../../hello-world/hello1/app.ts"],`

- If I instead revert `CodeUri` to point to the folder with package.json, then I need to indicate the path to the new function a different way. If I choose `Handler`, e.g. `Handler: hello1/app.lambdaHandler`, (assuming Handler evaluates from the `CodeUri` directory), then it appears to build correctly but `.aws-sam/build/HelloOneFunction` includes the wrong source, meaning the source for `HelloWorldFunction` instead. Additionally, the `app.js.map` file refers to a file in `private/var/folders` that does not exist:
```
  "sources": ["../../../../../../../../../private/var/folders/kw/x3x7dz_j07v_b98j6dy7hbx4blmgmx/T/tmpclbl29bx/app.ts"],
```
(everything up through the `/T` exists, but the `tmp...` folder does not.)

- If I instead refer to the path in `EntryPoints`, e.g. `hello2/app.ts`, then `.aws-sam/build/HelloTwoFunction/app.js` includes the correct source, but its `app.js.map` again points to a file in `private/var/folders` that does not exist.

If I leave those folders in place in the filesystem, but remove them from template.yaml, then I can still set breakpoints on the original `hello-world/app.ts` function. However, if I add `AWS::Serverless::Function` definitions for either of the two new subfolder functions into template.yaml, then the breakpoints don't work anymore - not even on the original `hello-world/app.ts` function.

I believe that for codebases with multiple functions, it would be preferable to share one package.json that is from a common parent folder, and to also be able to use a debugger on these functions.

**Comment by qingchm** [comment]
Hey guys this has been addressed in our most recent release 1.79.0 (https://github.com/aws/aws-sam-cli/releases/tag/v1.79.0). Closing the issue here. Thanks for all the engagement!
